### PR TITLE
faq for external clients

### PR DESCRIPTION
**Pull request by tbarnes-us**

### CODE REVIEW

File: docs-source/content/faq/external-clients.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+## WIP - This is a DRAFT!
+
+> __This document is an early draft.__
+> __Do not merge.__
+> __There is no need to review yet.__
+
+
+### Approaches
+
+There are at least three approaches for external WebLogic EJB or JMS client access to a Kubernetes hosted WebLogic cluster. These include [Federation](#federation), [Load Balancer Tunneling](#load-balancer-tunneling), and [Kubernetes Node Ports](#kubernetes-node-ports).
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: we do not support or recommend federation, we should not talk about it here

File: docs-source/content/faq/external-clients.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+- Deploy the remote clients inside another Kubernetes cluster.
+
+- Federate the two clusters so that they effectively share the same DNS namespace. See [Federation](https://kubernetes.io/docs/concepts/cluster-administration/federation/) in the Kubernetes documentation.  Note that federation is a relatively new Kubernetes feature as of 2019.
+
+- If the addresses (URLs) for remote access from the client cluster to the original cluster are not the same as the addresses used by clients running within the original cluster itself, then it is often necessary to configure a custom channel for the T3 protocol in the original cluster.  See [Adding a WebLogic Custom Channel](#adding-a-weblogic-custom-channel).
+
+
+#### Load Balancer Tunneling
+
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: There should be a sentence before the list to explain what the list is

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: Seems like you assume that people know what these are, and just jump right into setup instructions

File: docs-source/content/faq/external-clients.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+- Deploy a Kubernetes NodePort to publicly expose the WebLogic ports. See [Setting up a NodePort](#setting-up-a-nodeport).
+
+- Review the [Security Notes](#security-notes) below.
+
+### Adding a WebLogic Custom Channel
+
+#### When is a WebLogic Custom Channel needed?
+
+WebLogic implicitly creates a multi-protocol default network channel that spans the Listen Address and Port fields specified on each server in the cluster. Here are some reasons you may need to configure an additional dedicated WebLogic custom channel to handle remote EJB or JMS client network traffic:
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: I would not say "here are some reasons" - implies there are others, just say something like "you need to do x when:"

File: docs-source/content/faq/external-clients.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+- A default channel does not provide a way to configure an external address and port, but external listen address and/or port configuration is needed when a channel's configured listen address and/or port would not work if used to form a URL in the remote client.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: this wording is awkward, i would reword this. it reads like "you need to do x but you cant do x"

File: docs-source/content/faq/external-clients.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  - This is because remote EJB and JMS clients internally use their client's channel's configured network information to reconnect to WebLogic when needed. (The EJB and JMS clients do not always use the initial URL specified in the client's JNDI context.)
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: feels like this would be better as prose than a list

File: docs-source/content/faq/external-clients.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+- A default channel exposes multiple protocols but it's a best practice to limit the protocols that can be accessed by external clients.
+
+- A default channel may often be left insecure if it is only accessible internally, but it is often desirable to secure externally accessible channels (two-way SSL).
+
+#### Configuring a WebLogic Custom Channel
+
+The basic requirements for configuring a custom channel for remote EJB and JMS access are:
+
+- Configure a T3 protocol network-access-point (NAP) with the same name and same port on each server (the operator will set the listen address for you).
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: dont repeat "same"

File: docs-source/content/faq/external-clients.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+- Configure the external listen address and port on each NAP to match the address and port component of a URLs your clients can use. For example, if you are providing access to remote clients via a load balancer then these should match the address and port of the load balancer.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: via -> using

File: docs-source/content/faq/external-clients.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+- If you want WebLogic T3 clients to tunnel through HTTP, then enable HTTP tunneling on each NAP. This is often necessary for load balancers.
+
+- Do _NOT_ set `outbound-enabled` to `true` on the network access point (the default is `false`).
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: why?

File: docs-source/content/faq/external-clients.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+- Ensure you haven't enabled `calculated-listen-ports` for WebLogic dynamic cluster servers.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: explain why

File: docs-source/content/faq/external-clients.md [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+For example, here is a snippet of config.xml for channel 'MyChannel' defined for a WebLogic dynamic cluster named 'cluster-1':
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: put filenames and other technical names in backticks

File: docs-source/content/faq/external-clients.md [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+  set('Enabled', true)
+  set('TwoWaySslEnabled', false)
+  set('ClientCertificateEnforced', false)
+```
+
+In this example:
+
+- WebLogic binds the custom network channel to port 7999 and the default network channel to 8001.
+
+- The operator will automatically create a Kubernetes service named `DOMAIN_UID-cluster-cluster-1` for both the custom and default channel
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: either use periods at the end of every list item, or none of them - be consistent
@rosemarymarano can tell you what we normally do

File: docs-source/content/faq/external-clients.md [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+- Internal clients running in the same Kubernetes cluster as the channel can access the cluster using `t3://DOMAIN_UID-cluster-cluster-1:8001`.
+
+- External clients would be expected to access the cluster using the custom channel using URLs like `t3://some.public.address.com:30999` or, if leveraging tunneling, `http://some.public.address.com:30999`.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: leveraging -> using

File: docs-source/content/faq/external-clients.md [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+
+#### WebLogic Custom Channel Notes 
+
+- Channel configuration for a configured cluster requires configuring the same network-access-point on each server. The Operator currently doesn't test or support network channels that have a different configuration on each server in the cluster.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: be consistent - here you have "network-access-point" but earlier there were no hyphens

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: We do not capitalize "operator"

File: docs-source/content/faq/external-clients.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+- Additional steps are required for external clients beyond configuring the custom channel - see [Approaches](#approaches).
+
+### Setting up a NodePort
+
+#### Getting Started
+
+A Kubernetes NodePort exposes a port on each machine that hosts the Kubernetes cluster where the port is accessible from outside of a Kubernetes cluster. This port redirects network traffic to pods within the Kubernetes cluster. Setting up a Kubernetes NodePort is one of multiple approaches for giving external WebLogic clients access to JMS or EJBs (see [Approaches](#approaches) for alternatives).
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: remove "of multiple"

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: I don't think you need the back ref to approaches.. that was literally just above this right, they should have just read that :)

File: docs-source/content/faq/external-clients.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+- Additional steps are required for external clients beyond configuring the custom channel - see [Approaches](#approaches).
+
+### Setting up a NodePort
+
+#### Getting Started
+
+A Kubernetes NodePort exposes a port on each machine that hosts the Kubernetes cluster where the port is accessible from outside of a Kubernetes cluster. This port redirects network traffic to pods within the Kubernetes cluster. Setting up a Kubernetes NodePort is one of multiple approaches for giving external WebLogic clients access to JMS or EJBs (see [Approaches](#approaches) for alternatives).  
+
+If an EJB or JMS service is running on an Admin Server, then you can skip the rest of this section and use the `spec.adminServer.adminService.channels` domain resource attribute to have the Operator create a Node Port for you. See [Reference - Domain resource]({{<relref "/reference/domain-resource/_index.md">}}). On the other hand, if the EJB or JMS service is running in a WebLogic Cluster or stand-alone WebLogic managed server, and you desire to provide access to the service via a NodePort, then the NodePort must be deployed 'manually' - see the following sample and table.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: Don't say "on the other hand", if you have to say anything, use "Otherwise", or just start a new paragraph

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: I don't think we capitalize the "c" in "WebLogic cluster" - check with @rosemarymarano

File: docs-source/content/faq/external-clients.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+> __NOTE:__ Setting up a NodePort usually also requires setting up a custom network channel. See [Adding a WebLogic Custom Channel](#adding-a-weblogic-custom-channel) above.
+
+#### Sample NodePort Resource
+
+The following NodePort yaml deploys an external node port of 30999 and internal port 7999 for domain UID of DOMAIN_UID, a domain name of DOMAIN_NAME, and a cluster name of CLUSTER_NAME. It assumes that 7999 corresponds to a T3 protocol port of a channel that's configured on your WebLogic cluster.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: YAML

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: put the port numbers, names of things in backticks

Comment (markxnelson) [reply docs-source/content/faq/external-clients.md]: I think we say "DomainUID" not "domain UID" - check with @rosemarymarano

File: docs-source/content/faq/external-clients.md [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+|spec.ports.nodePort|The external port that clients will use. This must match the external port that's configured on the WebLogic configured channels/network-access-points. By default, Kubernetes requires that this value range from 30000 to 32767.|
+|spec.ports.port and spec.targetPort|These must match the port that's configured on the WebLogic configured channel/network-access-point(s)|
+
+### Security Notes
+
+- With some cloud providers, a load balancer or NodePort may implicitly expose a port to the public Internet. 
+
+- If such a port supports a protocol suitable for WebLogic clients, note that WebLogic allows access to JNDI entries, EJB/RMI applications, and JMS by anonymous users by default.
+
+- You can configure a custom channel with a secure protocol and two-way SSL to help prevent external access by unwanted clients.
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: probably should say how....

File: docs-source/content/faq/external-clients.md [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+
+
+### References
+
```
Comment (markxnelson) [inline docs-source/content/faq/external-clients.md]: put a lead in sentence here to explain why you are listing references.. are they just for further reading?

File: docs-source/content/faq/external-clients.md (hunk 15 above)
Comment (rosemarymarano) [reply docs-source/content/faq/external-clients.md]: We don't. WebLogic cluster is correct. And, standalone is one word.

File: docs-source/content/faq/external-clients.md (hunk 16 above)
Comment (rosemarymarano) [reply docs-source/content/faq/external-clients.md]: In code or as part of a path, we use `domainUID`. In text, we refer to domain UID.

File: docs-source/content/faq/external-clients.md (hunk 17 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: don't know how - i'm told setting up two-way SSL is a multi-step process

File: docs-source/content/faq/external-clients.md (hunk 16 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: I've  made the change for now.  Checking with Rosemary via slack.

File: docs-source/content/faq/external-clients.md (hunk 15 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: I've  made both changes for now.  Checking with Rosemary via slack about the "WebLogic cluster".

File: docs-source/content/faq/external-clients.md (hunk 14 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: Done.

File: docs-source/content/faq/external-clients.md (hunk 13 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: Done.

File: docs-source/content/faq/external-clients.md (hunk 12 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: Done.

File: docs-source/content/faq/external-clients.md (hunk 11 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: Fixed the missing period. Asking Rosemary about standard via slack.

File: docs-source/content/faq/external-clients.md (hunk 10 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: done

File: docs-source/content/faq/external-clients.md (hunk 8 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: done

File: docs-source/content/faq/external-clients.md (hunk 9 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: done

File: docs-source/content/faq/external-clients.md (hunk 7 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: ok

File: docs-source/content/faq/external-clients.md (hunk 6 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: ok

File: docs-source/content/faq/external-clients.md (hunk 5 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: ok - done

File: docs-source/content/faq/external-clients.md (hunk 4 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: attempted aa rewording

File: docs-source/content/faq/external-clients.md (hunk 3 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: attempted a rewording

File: docs-source/content/faq/external-clients.md (hunk 2 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: done

File: docs-source/content/faq/external-clients.md (hunk 1 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: removed

File: docs-source/content/faq/external-clients.md [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+date: 2019-11-21T21:23:03Z
+draft: false
+weight: 6
+---
+
+### Approaches
+
+There are two supported approaches for giving external WebLogic EJB or JMS clients access to a Kubernetes hosted WebLogic cluster: [Load Balancer Tunneling](#load-balancer-tunneling) and [Kubernetes Node Ports](#kubernetes-node-ports).
+
+> __NOTE:__ This FAQ is for remote EJB and JMS clients - not JTA clients. The operator does not currently support external WebLogic JTA access to a WebLogic cluster, as (a) external JTA access requires each server in the cluster to be individually addressable by the client, but this conflicts with (b) the current operator requirement that a network channel in a cluster have the same port across all servers in the cluster.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: For the Hugo UI, we use this code to surround NOTES text:
{{% notice note %}}
{{% /notice %}}

Also, I would suggest that you use the designation (a.) and (b.), because in the first instance, it looks like you are stating as (a) external... (instead of as an external ...,) but you are denoting neither. 
Or, as you've done in the Config Overrides doc, use the capital letters, A and B.
(Please apply this change globally,)

File: docs-source/content/faq/external-clients.md [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+
+#### Load Balancer Tunneling
+
+The Load Balancer Tunneling approach for giving external WebLogic EJB or JMS clients access to a Kubernetes hosted WebLogic cluster involves configuring a network channel on the desired WebLogic cluster that accepts `T3` protocol traffic that's tunnelled over HTTP, deploying a load balancer that redirects external HTTP network traffic to the desired WebLogic network channel, and ensuring tha EJB and JMS clients specify a URL that resolves the load balancer's network address.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: tha -> that

File: docs-source/content/faq/external-clients.md [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+
+#### Load Balancer Tunneling
+
+The Load Balancer Tunneling approach for giving external WebLogic EJB or JMS clients access to a Kubernetes hosted WebLogic cluster involves configuring a network channel on the desired WebLogic cluster that accepts `T3` protocol traffic that's tunnelled over HTTP, deploying a load balancer that redirects external HTTP network traffic to the desired WebLogic network channel, and ensuring tha EJB and JMS clients specify a URL that resolves the load balancer's network address.  
+
+Here are the specific steps:
+
+- Configure a custom channel for the T3 protocol in WebLogic that (a) enables HTTP Tunneling, and (b) specifies an external address and port that correspond to the address and port remote clients will use access the load balancer.  See  [Adding a WebLogic Custom Channel](#adding-a-weblogic-custom-channel) for samples and details.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: use access -> use to access

File: docs-source/content/faq/external-clients.md [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+
+- Setup a load balancer that redirects HTTP traffic to the custom channel. For a discussion of load balancers see [Ingress]({{<relref "/userguide/managing-domains/ingress/_index.md">}}). If you're also using OKE/OCI to host your Kubernetes cluster, also see [Using an OCI Load Balancer]({{<relref "weblogic-kubernetes-operator/faq/oci-lb/_index.md">}}).
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Setup -> Set up   (set up (v), setup (n) (adj))
For a discussion of load balancers see -> For a discussion of load balancers, (comma) see

File: docs-source/content/faq/external-clients.md [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+
+- __Important__: ensure that the load balancer configures the HTTP flow to be 'sticky' - for example, a Traefik load balancer has a `sticky sessions` option. This ensures that all of the packets of a tunneling client connection flow to the same pod, otherwise the connection will stall when its packets are load balanced to a different pod.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: If you want to use it, there is a Hugo UI designation for caution:
{{% notice warning %}}
{{% /notice %}}

If that's too strong, there are also designations for:
{{% notice info %}}
{{% /notice %}}
And
{{% notice tip %}}
{{% /notice %}}

Whatever you choose, please capitalize the first word, Ensure.

File: docs-source/content/faq/external-clients.md [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+
+- Remote clients can then access the custom channel using an 'http://' URL instead of a 't3://' URL.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: 'http://' -> `http://` (surround with back ticks)
't3://' -> `t3://`

File: docs-source/content/faq/external-clients.md [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+
+- Review the [Security Notes](#security-notes) below.
+
+#### Kubernetes Node Ports
+
+The Kubernetes Node Ports approach for giving external WebLogic EJB or JMS clients access to a Kubernetes hosted WebLogic cluster involves configuring a network channel on the desired WebLogic cluster that accepts `T3` protocol traffic, and deploying Kubernetes NodePort that redirects external network traffic on the Kubernetes nodes to the network channel.
+
+Here are the specific steps:
+
+- Configure a custom channel for the T3 protocol in WebLogic that specifies an external address and port that are suitable for remote client use.  See [Adding a WebLogic Custom Channel](#adding-a-weblogic-custom-channel).
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: T3 -> you've used it four times, twice in code font and twice in regular text. I would pick one (regular text is ok if it's not being used in code, as in a URL) and use it consistently.

File: docs-source/content/faq/external-clients.md [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+
+#### Configuring a WebLogic Custom Channel
+
+The basic requirements for configuring a custom channel for remote EJB and JMS access are:
+
+- Configure a T3 protocol network-access-point (NAP) with the same name and port on each server (the operator will set the listen address for you).
+
+- Configure the external listen address and port on each NAP to match the address and port component of a URLs your clients can use. For example, if you are providing access to remote clients using a load balancer then these should match the address and port of the load balancer.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: URLs -> URL

File: docs-source/content/faq/external-clients.md [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+    <maximum-dynamic-server-count>5</maximum-dynamic-server-count>
+    <calculated-listen-ports>false</calculated-listen-ports>
+    <server-name-prefix>managed-server</server-name-prefix>
+    <dynamic-cluster-size>5</dynamic-cluster-size>
+    <max-dynamic-cluster-size>5</max-dynamic-cluster-size>
+  </dynamic-servers>
+</cluster>
+```
+
+And here is a snippet of off-line WLST that corresponds to the above config.xml snippet:
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: off-line WLST -> offline WLST code

File: docs-source/content/faq/external-clients.md [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+
+- Channel configuration for a configured cluster requires configuring the same network-access-point on each server. The operator currently doesn't test or support network channels that have a different configuration on each server in the cluster.
+
+- Additional steps are required for external clients beyond configuring the custom channel - see [Approaches](#approaches).
+
+### Setting up a NodePort
+
+#### Getting Started
+
+A Kubernetes NodePort exposes a port on each machine that hosts the Kubernetes cluster where the port is accessible from outside of a Kubernetes cluster. This port redirects network traffic to pods within the Kubernetes cluster. Setting up a Kubernetes NodePort is one approaches for giving external WebLogic clients access to JMS or EJBs.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: one approaches -> one approach

File: docs-source/content/faq/external-clients.md [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+
+If an EJB or JMS service is running on an Admin Server, then you can skip the rest of this section and use the `spec.adminServer.adminService.channels` domain resource attribute to have the operator create a Node Port for you. See [Reference - Domain resource]({{<relref "/reference/domain-resource/_index.md">}}). Otherwise, if the EJB or JMS service is running in a WebLogic cluster or stand-alone WebLogic managed server, and you desire to provide access to the service using a NodePort, then the NodePort must be deployed 'manually' - see the following sample and table.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Admin Server -> Administration Server
stand-alone -> standalone
managed server -> Managed Server
Node Port - not sure if it should be one word (as in code) or two words, but, unless there is a reason to write it differently, it's best to stick to one designation. (In the operator docs, particularly when it is capitalized as such, NodePort, it's usually written as one word.)

File: docs-source/content/faq/external-clients.md [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+
+> __NOTE:__ Setting up a NodePort usually also requires setting up a custom network channel. See [Adding a WebLogic Custom Channel](#adding-a-weblogic-custom-channel) above.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Your NOTE designation appears faded and hard to read. I suggest using the Hugo UI code text:
{{% notice note %}}
{{% /notice %}}

File: docs-source/content/faq/external-clients.md [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+    port: 7999
+    protocol: TCP
+    targetPort: 7999
+```
+
+#### Table of NodePort Attributes
+
+|Attribute|Description|
+|---------|-----------|
+|metadata.name|For this particular use case, the NodePort name can be arbitrary as long as it is DNS compatible. But, as a convention it's recommended to use `DOMAIN_UID-cluster-CLUSTER_NAME-ext`. To ensure the name is DNS compatible, use all lower case and convert any underscores (`_`) to dashes (`-`).|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: I suggest using back ticks (code font) for all the attribute names.

File: docs-source/content/faq/external-clients.md [hunk 32]
```diff
@@ -0,0 +1,202 @@
...
+|metadata.namespace|Must match the namespace of your WebLogic cluster.|
+|metadata.labels|Optional. It's helpful to set a weblogic.domainUid so that cleanup scripts can locate all Kubernetes resources associated with a particular domain UID.|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: weblogic.domainUid -> `weblogic.domainUid`

File: docs-source/content/faq/external-clients.md [hunk 33]
```diff
@@ -0,0 +1,202 @@
...
+|spec.type|Must be `NodePort`|
+|spec.externalTrafficPolicy|Set to `Cluster` for most use cases. This may lower performance, but ensures that a client that attaches to a Node without any pods that match the `spec.selector` will be rerouted to a node with pods that do match. If set to `Local`, then connections to a particular Node will only route to that Node's pods and will fail if the Node doesn't host any pods with the given `spec.selector`. It's recommended for clients of a `spec.externalTrafficPolicy: Local` NodePort to use a URL that resolves to a list of all nodes such as `t3://mynode1,mynode2:30999` so that a client connect attempt will implicitly try mynode2 if mynode1 fails (alternatively, use a round-robin DNS address in place of `mynode1,mynode2`).|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: No reason to capitalize Node in this description text.
mynode1 -> `mynode1`
mynode2 -> `mynode2`

File: docs-source/content/faq/external-clients.md [hunk 34]
```diff
@@ -0,0 +1,202 @@
...
+|spec.sessionAffinity|Set to `ClientIP` to ensure an HTTP tunneling connection always routes to the same pod, otherwise the connection may hang and fail.|
+|spec.selector|Specify a weblogic.domainUID and weblogic.clusterName to associate the NodePort resource with your cluster's pods. The operator automatically sets these labels on the WebLogic cluster pods that it deploys for you.|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Specify -> Specifies
weblogic.domainUID -> `weblogic.domainUID`
weblogic.clusterName -> `weblogic.clusterName`

File: docs-source/content/faq/external-clients.md [hunk 35]
```diff
@@ -0,0 +1,202 @@
...
+|spec.ports.name|This name is arbitrary.|
+|spec.ports.nodePort|The external port that clients will use. This must match the external port that's configured on the WebLogic configured channels/network-access-points. By default, Kubernetes requires that this value range from `30000` to `32767`.|
+|spec.ports.port and spec.targetPort|These must match the port that's configured on the WebLogic configured channel/network-access-point(s)|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Add a period at the end of the sentence.

File: docs-source/content/faq/external-clients.md [hunk 36]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+#### Table of NodePort Attributes
+
+|Attribute|Description|
+|---------|-----------|
+|metadata.name|For this particular use case, the NodePort name can be arbitrary as long as it is DNS compatible. But, as a convention it's recommended to use `DOMAIN_UID-cluster-CLUSTER_NAME-ext`. To ensure the name is DNS compatible, use all lower case and convert any underscores (`_`) to dashes (`-`).|
+|metadata.namespace|Must match the namespace of your WebLogic cluster.|
+|metadata.labels|Optional. It's helpful to set a weblogic.domainUid so that cleanup scripts can locate all Kubernetes resources associated with a particular domain UID.|
+|spec.type|Must be `NodePort`|
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: Add a period at the end of the sentence.

File: docs-source/content/faq/external-clients.md [hunk 37]
```diff
@@ -0,0 +1,202 @@
...
+- If such a port supports a protocol suitable for WebLogic clients, note that WebLogic allows access to JNDI entries, EJB/RMI applications, and JMS by anonymous users by default.
+
+- You can configure a custom channel with a secure protocol and two-way SSL to help prevent external access by unwanted clients. See [When is a WebLogic Custom Channel needed?](#when-is-a-weblogic-custom-channel-needed?).
+
+
+### Optional Reading
+
+- See [Run Standalone WebLogic JMS Clients on Kubernetes](https://blogs.oracle.com/weblogicserver/run-standalone-weblogic-jms-clients-on-kubernetes) for sample JMS client code and JMS configuration.
+
+- See [T3 RMI Communication for WebLogic Server Running on Kubernetes](https://blogs.oracle.com/weblogicserver/t3-rmi-communication-for-weblogic-server-running-on-kubernetes) for a deep level discussion of using T3 in combination with port mapping.
```
Comment (rosemarymarano) [inline docs-source/content/faq/external-clients.md]: deep level -> deep-level

File: docs-source/content/faq/external-clients.md (hunk 25 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: OK. I switched to unadorned T3 throughout.

File: docs-source/content/faq/external-clients.md (hunk 23 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: I chose to leave it as is, but with capitalized 'Ensure'.  It looks more clear to me when the 'sticky' instruction step in-lined with the other steps.

File: docs-source/content/faq/external-clients.md (hunk 17 above)
Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: can we defer and file a JIRA to add a 'two-way SSL' FAQ entry?

Comment (tbarnes-us) [reply docs-source/content/faq/external-clients.md]: punting this after further discussion with Mark
